src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -90,62 +105,119 @@ public ShooterSubsystem(ShooterHoodSubsystem shooterHood) {
+    } else {
+      leader_shooterMAXRight.set(MoPrefs.getShooterFlywheelSetpoint());
+    }
+
+    final boolean shooterHoodReady = shooterHood.hasReliableZero() && shooterHood.getFullyDeployed();
+    final boolean shooterWheelReady = MoPrefs.getShooterFlywheelSetpoint()
+        - leader_shooterMAXRight.getEncoder().getVelocity() < MoPrefs.getShooterFlywheelTolerance();
+    if (shooterHoodReady && shooterWheelReady) {
+      shooterGate.set(MoPrefs.getShooterGateSetpoint());
+    } else {
+      shooterGate.set(0);
+    }
+
+  }
+
+  public void shootAuto() {
+    // NetworkTableInstance.getDefault().getTable("limelight").getEntry("ledMode").setNumber(3);
     shooterHood.deployHood();
-    shooterPIDRight.setReference(MoPrefs.getShooterFlywheelSetpoint(), ControlType.kVelocity);
-    if (shooterHood.getFullyDeployed()
-        && MoPrefs.getShooterFlywheelSetpoint() - shooterMAXRight.getEncoder().getVelocity() < 0.1) {
+
+    leader_shooterMAXRight.set(MoPrefs.getShooterFlywheelSetpoint());
+
+    if (shooterHood.hasReliableZero() && shooterHood.getFullyDeployed() && MoPrefs.getShooterFlywheelSetpoint()
+        - leader_shooterMAXRight.getEncoder().getVelocity() < MoPrefs.getShooterFlywheelTolerance()) {
       shooterGate.set(MoPrefs.getShooterGateSetpoint());
     } else {
       shooterGate.set(0);
     }
   }
 
   public void idle() {
-    // stow hood
     // stop gate
     // slow shooter wheel
-    shooterHood.stowHood();
-    shooterPIDRight.setReference(MoPrefs.getShooterFlywheelIdle(), ControlType.kVelocity);
     shooterGate.stopMotor();
+    // NetworkTableInstance.getDefault().getTable("limelight").getEntry("ledMode").setNumber(1);
+
+    if (maintainFlywheelAtIdle) {
+      if (enablePID) {
+        shooterPIDRight.setReference(MoPrefs.getShooterFlywheelIdle(), ControlType.kVelocity);
+      } else {
+        leader_shooterMAXRight.set(MoPrefs.getShooterFlywheelIdle());
+      }
+    } else {
+      if (enablePID) {
+        shooterPIDRight.setReference(0, ControlType.kVelocity);
+      } else {
+        leader_shooterMAXRight.stopMotor();
+      }
+    }
   }
 
   public void purge() {
     // stow hood
     // reverse gate
     // reverse shooter wheel
-    shooterHood.stowHood();
-    shooterPIDRight.setReference(-MoPrefs.getShooterFlywheelIdle(), ControlType.kVelocity);
-    shooterGate.set(-MoPrefs.getShooterGateSetpoint());
+    if (enablePID) {
+      shooterPIDRight.setReference(-1 * MoPrefs.getShooterFlywheelIdle(), ControlType.kVelocity);
+    } else {
+      leader_shooterMAXRight.set(-1 * MoPrefs.getShooterFlywheelIdle());
+    }
+
+    shooterGate.set(-1 * MoPrefs.getShooterGateSetpoint());
+  }
+
+  private void updateVelocity() {
+    double currPos = shooterEncoder.getPosition();
+    double currTime = Timer.getFPGATimestamp();
+    double dPos = currPos - lastPos;
+    double dT = currTime - lastTime;
+    lastPos = currPos;
+    lastTime = currTime;
+    currVelocity = dPos / dT * 60;
```

Comment:
If this continues to be noisy, might need to add a median filter to reject bad samples.